README.md
# Reveal Template

A plugin for WordPress to reveal the theme template file used to render the displayed page, via the admin bar, footer, widget, shortcode, and/or template tag.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/reveal-template/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Optionally customize the plugin's settings by clicking the plugin's 'Settings' link next to its 'Deactivate' link (still on the Plugins page), or click on the 'Design' -> 'Reveal Template' link, to go to the plugin's admin settings page. Or use the 'Reveal Template' widget. Or use the shortcode.


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/reveal-template/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/reveal-template/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/reveal-template/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/reveal-template/reviews/) for it in the WordPress Plugin Directory or considerating a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 3.7 _(2021-11-14)_

### Highlights:

This minor release updates the plugin framework, notes compatibility through WP 5.8+, and reorganizes unit tests.

### Details:

* Change: Note compatibility through WP 5.8+
* Change: Update plugin framework to 064
    * 064:
    * New: For checkbox settings, support a 'more_help' config option for defining help text to appear below checkbox and its label
    * Fix: Fix URL for plugin listing donate link
    * Change: Store donation URL as object variable
    * Change: Update strings used for settings page donation link
    * 063:
    * Fix: Simplify settings initialization to prevent conflicts with other plugins
    * Change: Remove ability to detect plugin settings page before current screen is set, as it is no longer needed
    * Change: Enqueue thickbox during `'admin_enqueue_scripts'` action instead of during `'init'`
    * Change: Use `is_plugin_admin_page()` in `help_tabs()` instead of reproducing its functionality
    * Change: Trigger a debugging warning if `is_plugin_admin_page()` is used before `'admin_init'` action is fired
    * 062:
    * Change: Update `is_plugin_admin_page()` to use `get_current_screen()` when available
    * Change: Actually prevent object cloning and unserialization by throwing an error
    * Change: Check that there is a current screen before attempting to access its property
    * Change: Remove 'type' attribute from `style` tag
    * Change: Incorporate commonly defined styling for inline_textarea
* Change: Tweak installation instruction
* Unit tests:
    * Change: Restructure unit test file structure
        * Change: Move `tests/` into `tests/phpunit/tests/`
        * Change: Move `tests/bootstrap.php` to `tests/phpunit/`
        * Change: Move `bin/` into `phpunit/`
        * Change: In bootstrap, store path to plugin file constant so its value can be used within that file and in test file
        * Change: In bootstrap, check for test installation in more places and exit with error message if not found
        * Change: Remove 'test-' prefix from unit test files
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Tweak some formatting in TODO.md
* New: Add a few more possible TODO items

## 3.6.1 _(2021-04-14)_
* Fix: Update plugin framework to 061 to fix a bug preventing settings from getting saved.

## 3.6 _(2021-04-01)_

### Highlights:

This minor release adds support for the 'privacypolicy' template, updates the plugin framework, improves some inline and UI help text, and notes compatibility through WP 5.7+.

### Details:

* New: Add support for 'privacypolicy' template
* Change: Update plugin framework to 060
    * 060:
    * Rename class from `c2c_{PluginName}_Plugin_051` to `c2c_Plugin_060`
    * Move string translation handling into inheriting class making the plugin framework code plugin-agnostic
        * Add abstract function `get_c2c_string()` as a getter for translated strings
        * Replace all existing string usage with calls to `get_c2c_string()`
    * Handle WordPress's deprecation of the use of the term "whitelist"
        * Change: Rename `whitelist_options()` to `allowed_options()`
        * Change: Use `add_allowed_options()` instead of deprecated `add_option_whitelist()` for WP 5.5+
        * Change: Hook `allowed_options` filter instead of deprecated `whitelist_options` for WP 5.5+
    * New: Add initial unit tests (currently just covering `is_wp_version_cmp()` and `get_c2c_string()`)
    * Add `is_wp_version_cmp()` as a utility to compare current WP version against a given WP version
    * Refactor `contextual_help()` to be easier to read, and correct function docblocks
    * Don't translate urlencoded donation email body text
    * Add inline comments for translators to clarify purpose of placeholders
    * Change PHP package name (make it singular)
    * Tweak inline function description
    * Note compatibility through WP 5.7+
    * Update copyright date (2021)
    * 051:
    * Allow setting integer input value to include commas
    * Use `number_format_i18n()` to format integer value within input field
    * Update link to coffee2code.com to be HTTPS
    * Update `readme_url()` to refer to plugin's readme.txt on plugins.svn.wordpress.org
    * Remove defunct line of code
* Change: Move translation of all parent class strings into main plugin file
* Hardening: Escape URLs before being displayed
* Change: Include mention of the template also being revealed in the admin bar by default
* Change: Refactor how setting page explanatory text is output
* Change: Tweak help text for display_in_footer setting to reflect additional display methods
* Change: Fix typos in readme
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)

## 3.5 _(2020-07-18)_

### Highlights:

This feature release adds an admin bar entry to reveal the current template, updates the plugin framework, adds a TODO.md file, updates a few URLs to be HTTPS, expands unit testing, and updates compatibility to be WP 4.9-5.4+.

### Details:

* New: Add integration with admin bar
    * New: Add admin bar entry for revealing template
    * New: Add setting to control if admin bar entry should appear (defaulted to true)
    * New: Add `can_show_in_admin_bar()` for determining if the admin bar item should be shown
    * New: Add new screenshot
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add items to it)
* Fix: Correct typo on plugin settings page
* Change: Update plugin framework to 050
    * 050:
    * Allow a hash entry to literally have '0' as a value without being entirely omitted when saved
    * Output donation markup using `printf()` rather than using string concatenation
    * Update copyright date (2020)
    * Note compatibility through WP 5.4+

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Reveal_Template
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/reveal-template.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/test-c2c-plugin.php
<?php

defined( 'ABSPATH' ) or die();

class c2c_Plugin extends WP_UnitTestCase {

	protected $obj;

	public function setUp() {
		parent::setUp();

		add_filter( 'gettext_reveal-template', array( $this, 'translate_text' ), 10, 2 );
		$this->obj = c2c_RevealTemplate::get_instance();
	}

	public function tearDown() {
		parent::tearDown();
	}


	//
	//
	// HELPERS
	//
	//


	public function translate_text( $translation, $text ) {
		if ( 'Donate' === $text ) {
			$translation = 'Donar';
		}

		return $translation;
	}


	//
	//
	// DATA PROVIDERS
	//
	//


	public static function wp_version_comparisons() {
		return array(
			//[ WP ver, version to compare to, operator, expected result ]
			[ '5.5', '5.5.1', '>=', false ],
			[ '5.5', '5.6',   '>=', false ],
			[ '5.5', '5.5',   '>=', true ],
			[ '5.5', '5.4.3', '>=', true ],
			[ '5.5', '5.5.1', '',   false ],
			[ '5.5', '5.6',   '',   false ],
			[ '5.5', '5.5',   '',   true ],
			[ '5.5', '5.5',   '>',  false ],
			[ '5.5', '5.5',   '<',  false ],
			[ '5.5', '5.5.1', '<=', true ],
			[ '5.5', '5.6',   '<=', true ],
			[ '5.5', '5.5',   '<=', true ],
			[ '5.5', '5.4.3', '<=', false ],
			[ '5.5', '5.5',   '=',  true ],
			[ '5.5', '5.5.1', '=',  false ],
			[ '5.5', '5.5',   '!=', false ],
		);
	}


	//
	//
	// TESTS
	//
	//


	/*
	 * is_wp_version_cmp()
	 */

	/**
	 * @dataProvider wp_version_comparisons
	 */
	public function test_is_wp_version_cmp( $wp_ver, $ver, $op, $expected ) {
		global $wp_version;
		$orig_wp_verion = $wp_version;

		$wp_version = $wp_ver;
		$this->{ $expected ? 'assertTrue' : 'assertFalse' }( $this->obj->is_wp_version_cmp( $ver, $op ) );

		$wp_version = $orig_wp_verion;
	}

	/*
	 * get_c2c_string()
	 */

	 public function test_get_c2c_string_size() {
		$this->assertEquals( 21, count( $this->obj->get_c2c_string( '' ) ) );
	}

	public function test_get_c2c_string_for_unknown_string() {
		$str = 'unknown string';

		$this->assertEquals( $str, $this->obj->get_c2c_string( $str ) );
	}

	public function test_get_c2c_string_for_known_string_translated() {
		$this->assertEquals( 'Donar', $this->obj->get_c2c_string( 'Donate' ) );
	}

	public function test_get_c2c_string_for_known_string_untranslated() {
		$str = 'A value is required for: "%s"';

		$this->assertEquals( $str, $this->obj->get_c2c_string( $str ) );
	}

}


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Reveal_Template
 */

define( 'REVEAL_TEMPLATE_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/reveal-template.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require REVEAL_TEMPLATE_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


reveal-template.widget.php
<?php
/**
 * Reveal Widget plugin widget code.
 *
 * Copyright (c) 2013-2020 by Scott Reilly (aka coffee2code)
 *
 * @package c2c_RevealTemplateWidget
 * @author  Scott Reilly
 * @version 004
 */

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_RevealTemplateWidget' ) ) :

require_once( dirname( __FILE__ ) . DIRECTORY_SEPARATOR . 'c2c-widget.php' );

class c2c_RevealTemplateWidget extends c2c_Widget_013 {

	protected static $template_path_types;

	/**
	 * Returns version of the widget.
	 *
	 * @since 004
	 *
	 * @return string
	 */
	public static function version() {
		return '004';
	}

	/**
	 * Registers the widget.
	 *
	 * @since 003
	 */
	public static function register_widget() {
		register_widget( __CLASS__ );
	}

	/**
	 * Constructor.
	 */
	public function __construct() {
		// Load textdomain.
		load_plugin_textdomain( 'reveal-template' );

		parent::__construct( 'reveal-template', __FILE__, array() );
	}

	/**
	 * Initializes the plugin's configuration and localizable text variables.
	 */
	public function load_config() {
		$this->title       = __( 'Reveal Template', 'reveal-template' );
		$this->description = __( 'Reveal the name of the theme template used to render the displayed page.', 'reveal-template' );

		$this->config = array(
			'title' => array(
				'input'   => 'text',
				'default' => __( 'Revealed Template', 'reveal-template' ),
				'label'   => __( 'Title', 'reveal-template' ),
			),
			'template_path_type' => array(
				'input'   => 'select',
				'default' => c2c_RevealTemplate::get_instance()->get_default_template_path_type(),
				'label'   => __( 'Template path type', 'reveal-template' ),
				'options' => array_keys( $this->get_template_path_types() ),
				'help'    => $this->get_template_path_types_help(),
			),
			'show_non_admins' => array( 'input' => 'checkbox',
				'label'   => __( 'Show widget to all visitors?', 'reveal-template' ),
				'help'    => __( 'If checked, the widget will always be visible. By default (and when unchecked), the widget is only shown to users with the "update_themes" capability, which is generally only admins.', 'reveal-template' ),
			),
		);
	}

	/**
	 * Returns the template path types and their descriptions.
	 *
	 * @return array
	 */
	private function get_template_path_types() {
		if ( ! self::$template_path_types ) {
			self::$template_path_types = c2c_RevealTemplate::get_instance()->get_template_path_types();
		}

		return self::$template_path_types;
	}

	/**
	 * Returns string for help text for template path types setting.
	 *
	 * @return string
	 */
	private function get_template_path_types_help() {
		$help = '';
		foreach ( $this->get_template_path_types() as $key => $text ) {
			$help .= "<strong>$key</strong>: $text<br />";
		}
		return $help;
	}

	/**
	 * Outputs the body of the widget.
	 *
	 * @param array   $args     Widget args.
	 * @param array   $instance Widget instance.
	 * @param array   $settings Widget settings.
	 * @return string The widget body content.
	 */
	public function widget_body( $args, $instance, $settings ) {
		extract( $args );
		extract( $settings );

		$admin_only = ! isset( $show_non_admins ) || empty( $show_non_admins ) || '0' === $show_non_admins;
		$args = array(
			'admin_only' => $admin_only, // Abide by widget setting
			'echo'       => false, // Never echo
			'return'     => false, // Only get a return value is user is able to see value
		);

		return c2c_RevealTemplate::get_instance()->reveal( $template_path_type, $args );
	}

	/**
	 * Validates widget instance values.
	 *
	 * @param array  $instance Array of widget instance values
	 * @return array The filtered array of widget instance values
	 */
	public function validate( $instance ) {
		if ( ! in_array( $instance['template_path_type'], array( 'absolute', 'relative', 'theme-relative', 'filename' ) ) ) {
			$instance['template_path_type'] = 'absolute';
		}
		return $instance;
	}

} // end class

add_action( 'widgets_init', array( 'c2c_RevealTemplateWidget', 'register_widget' ) );

endif;


reveal-template.php
<?php
/**
 * Plugin Name: Reveal Template
 * Version:     3.7
 * Plugin URI:  https://coffee2code.com/wp-plugins/reveal-template/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Text Domain: reveal-template
 * Description: Reveal the theme template file used to render the displayed page, via the footer, widget, shortcode, and/or template tag.
 *
 * Compatible with WordPress 4.9+ through 5.8+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/reveal-template/
 *
 * @package Reveal_Template
 * @author  Scott Reilly
 * @version 3.7
 */

/*
	Copyright (c) 2008-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_RevealTemplate' ) ) :

require_once( dirname( __FILE__ ) . DIRECTORY_SEPARATOR . 'c2c-plugin.php' );
require_once( dirname( __FILE__ ) . DIRECTORY_SEPARATOR . 'reveal-template.widget.php' );

final class c2c_RevealTemplate extends c2c_Plugin_064 {

	/**
	 * Name of plugin's setting.
	 *
	 * @since 3.4
	 * @var string
	 */
	const SETTING_NAME = 'c2c_reveal_template';

	/**
	 * The one true instance.
	 *
	 * @var c2c_RevealTemplate
	 * @since 3.0
	 */
	private static $instance;

	/**
	 * The default template path type.
	 *
	 * @var string
	 * @since 3.0
	 */
	private $default_template_path_type = 'theme-relative';

	/**
	 * The shortcode name.
	 *
	 * @var string
	 * @since 3.0
	 */
	private $shortcode = 'revealtemplate';

	/**
	 * The template being used.
	 *
	 * @var string
	 */
	private $template = '';

	/**
	 * The memoized template path types array.
	 *
	 * @var array
	 * @since 3.0
	 */
	private static $template_path_types;

	/**
	 * Get singleton instance.
	 *
	 * @since 3.0
	 */
	public static function get_instance() {
		if ( ! isset( self::$instance ) ) {
			self::$instance = new self();
		}

		return self::$instance;
	}

	/**
	 * Constructor.
	 */
	protected function __construct() {
		parent::__construct( '3.7', 'reveal-template', 'c2c', __FILE__, array( 'settings_page' => 'themes' ) );
		register_activation_hook( __FILE__, array( __CLASS__, 'activation' ) );

		return self::$instance = $this;
	}

	/**
	 * Handles activation tasks, such as registering the uninstall hook.
	 *
	 * @since 2.1
	 */
	public static function activation() {
		register_uninstall_hook( __FILE__, array( __CLASS__, 'uninstall' ) );
	}

	/**
	 * Handles uninstallation tasks, such as deleting plugin options.
	 *
	 * @since 2.1
	 */
	public static function uninstall() {
		delete_option( self::SETTING_NAME );
	}

	/**
	 * Returns translated strings used by c2c_Plugin parent class.
	 *
	 * @since 3.6
	 *
	 * @param string $string Optional. The string whose translation should be
	 *                       returned, or an empty string to return all strings.
	 *                       Default ''.
	 * @return string|string[] The translated string, or if a string was provided
	 *                         but a translation was not found then the original
	 *                         string, or an array of all strings if $string is ''.
	 */
	public function get_c2c_string( $string = '' ) {
		$strings = array(
			'A value is required for: "%s"'
				/* translators: %s: Label for setting. */
				=> __( 'A value is required for: "%s"', 'reveal-template' ),
			'Click for more help on this plugin'
				=> __( 'Click for more help on this plugin', 'reveal-template' ),
			' (especially check out the "Other Notes" tab, if present)'
				=> __( ' (especially check out the "Other Notes" tab, if present)', 'reveal-template' ),
			'Coffee fuels my coding.'
				=> __( 'Coffee fuels my coding.', 'reveal-template' ),
			'Did you find this plugin useful?'
				=> __( 'Did you find this plugin useful?', 'reveal-template' ),
			'Donate'
				=> __( 'Donate', 'reveal-template' ),
			'Expected integer value for: %s'
				=> __( 'Expected integer value for: %s', 'reveal-template' ),
			'Invalid file specified for C2C_Plugin: %s'
				/* translators: %s: Path to the plugin file. */
				=> __( 'Invalid file specified for C2C_Plugin: %s', 'reveal-template' ),
			'More information about %1$s %2$s'
				/* translators: 1: plugin name 2: plugin version */
				=> __( 'More information about %1$s %2$s', 'reveal-template' ),
			'More Help'
				=> __( 'More Help', 'reveal-template' ),
			'More Plugin Help'
				=> __( 'More Plugin Help', 'reveal-template' ),
			'Please consider a donation'
				=> __( 'Please consider a donation', 'reveal-template' ),
			'Reset Settings'
				=> __( 'Reset Settings', 'reveal-template' ),
			'Save Changes'
				=> __( 'Save Changes', 'reveal-template' ),
			'See the "Help" link to the top-right of the page for more help.'
				=> __( 'See the "Help" link to the top-right of the page for more help.', 'reveal-template' ),
			'Settings'
				=> __( 'Settings', 'reveal-template' ),
			'Settings reset.'
				=> __( 'Settings reset.', 'reveal-template' ),
			'Something went wrong.'
				=> __( 'Something went wrong.', 'reveal-template' ),
			'The plugin author homepage.'
				=> __( 'The plugin author homepage.', 'reveal-template' ),
			"The plugin configuration option '%s' must be supplied."
				/* translators: %s: The setting configuration key name. */
				=>__( "The plugin configuration option '%s' must be supplied.", 'reveal-template' ),
			'This plugin brought to you by %s.'
				/* translators: %s: Link to plugin author's homepage. */
				=> __( 'This plugin brought to you by %s.', 'reveal-template' ),
		);

		if ( ! $string ) {
			return array_values( $strings );
		}

		return ! empty( $strings[ $string ] ) ? $strings[ $string ] : $string;
	}

	/**
	 * Initializes the plugin's configuration and localizable text variables.
	 */
	public function load_config() {
		$this->name      = __( 'Reveal Template', 'reveal-template' );
		$this->menu_name = $this->name;

		$this->config = array(
			'display_in_footer' => array(
				'input'    => 'checkbox',
				'default'  => true,
				'label'    => __( 'Reveal in footer?', 'reveal-template' ),
				'help'     => __( 'To be precise, this displays where <code>&lt;?php wp_footer(); ?></code> is called. Refer to the documentation for additional ways to display the template (e.g. widget, shortcode, template tag).', 'reveal-template' ),
			),
			'display_in_admin_bar' => array(
				'input'    => 'checkbox',
				'default'  => true,
				'label'    => __( 'Reveal in admin bar?', 'reveal-template' ),
				'help'     => __( 'Adds a "Reveal Template" admin bar entry that only appears on the front-end of the site.', 'reveal-template' ),
			),
			'format' => array(
				'input'    => 'long_text',
				'default'  => __( '<p>Rendered template: %template%</p>', 'reveal-template' ),
				'label'    => __( 'Output format', 'reveal-template' ),
				'required' => true,
				'help'     => __( 'Only used for the footer display. Use %template% to indicate where the template name should go.', 'reveal-template' ),
			),
			'template_path' => array(
				'input'    => 'select',
				'datatype' => 'hash',
				'default'  => $this->get_default_template_path_type(),
				'label'    => __( 'Template path', 'reveal-template' ),
				'options'  => self::get_template_path_types(),
				'help'     => __( 'How much of the template path do you want reported? Applies directly to admin bar and footer displays, and is the default for the template tag usage (though can be overridden via an argument to <code>c2c_reveal_template()</code>)', 'reveal-template' ),
			),
		);
	}

	/**
	 * Override the plugin framework's register_filters() to actually register
	 * actions against filters.
	 *
	 * @link https://developer.wordpress.org/reference/functions/get_query_template/
	 */
	public function register_filters() {
		$options = $this->get_options();

		$templates = array(
			'404',
			'archive',
			'attachment',
			'author',
			'category',
			'date',
			'embed',
			'frontpage',
			'home',
			'index',
			'page',
			'paged',
			'privacypolicy',
			'search',
			'single',
			'singular',
			'tag',
			'taxonomy',
		);

		foreach ( $templates as $template ) {
			add_filter( $template . '_template', array( $this, 'template_handler' ) );
		}

		if ( $options['display_in_footer'] ) {
			add_action( 'wp_footer', array( $this, 'reveal_in_footer' ) );
		}

		if ( $this->can_show_in_admin_bar() ) {
			add_action( 'wp_before_admin_bar_render', array( $this, 'output_admin_bar_styles' ) );
			add_action( 'admin_bar_menu', array( $this, 'add_to_admin_bar' ), 100 );
		}

		add_shortcode( $this->shortcode, array( $this, 'shortcode' ) );
	}

	/**
	 * Outputs the text above the setting form.
	 *
	 * @param string $localized_heading_text Optional. Localized page heading text.
	 */
	public function options_page_description( $localized_heading_text = '' ) {
		$options = $this->get_options();
		parent::options_page_description( __( 'Reveal Template Settings', 'reveal-template' ) );
		?>

		<p>
			<?php _e( 'Reveal the theme template used to render the displayed page. By default this appears in the admin bar and the site\'s footer, but only for logged in users with the "update_themes" capability (such as an admin).', 'reveal-template' ); ?>
		</p>
		<p>
			<?php printf(
				__( 'Also note that the plugin provides a "Reveal Template" <a href="%s">widget</a> that can be used to reveal the current template.', 'reveal-template' ),
				esc_url( admin_url( 'widgets.php' ) )
			); ?>
		</p>
		<p>
			<?php printf(
				__( 'Please refer to this plugin\'s <a href="%s" title="readme">readme.txt</a> file for documentation and examples.', 'reveal-template' ),
				esc_url( $this->readme_url() )
			); ?>
		</p>

		<?php
	}

	/**
	 * Stores the name of the template being rendered.
	 *
	 * @param string  $template The template name.
	 * @return string The unmodified template name.
	 */
	public function template_handler( $template ) {
		$this->template = $template;
		return $template;
	}

	/**
	 * Returns types of, and descriptions for, the valid template path types.
	 *
	 * @since 3.0
	 *
	 * @return array Keys are the template path types, values are the translated descriptions
	 */
	public function get_template_path_types() {
		if ( ! self::$template_path_types ) {
			self::$template_path_types = array(
				'absolute'       => __( 'Absolute path, e.g. /usr/local/www/yoursite/wp-content/themes/yourtheme/single.php', 'reveal-template' ),
				'relative'       => __( 'Relative path, e.g. wp-content/themes/yourtheme/single.php', 'reveal-template' ),
				'theme-relative' => __( 'Path relative to themes directory, e.g. yourtheme/single.php', 'reveal-template' ),
				'filename'       => __( 'Filename, e.g. single.php', 'reveal-template' ),
			);
		}

		return self::$template_path_types;
	}

	/**
	 * Gets the default template path type.
	 *
	 * @since 3.0
	 *
	 * @return string
	 */
	public function get_default_template_path_type() {
		return $this->default_template_path_type;
	}

	/**
	 * Determines if the current user can be shown the template name/path.
	 *
	 * @since 3.0
	 */
	public function reveal_to_current_user() {
		return current_user_can( 'update_themes' );
	}

	/**
	 * Handles the shortcode.
	 *
	 * @since 3.0
	 *
	 * @param array  $atts    The shortcode attributes parsed into an array.
	 * @param string $content The content between opening and closing shortcode tags.
	 * @return string
	 */
	public function shortcode( $atts, $content = null ) {
		$defaults = array(
			'type'  => $this->get_default_template_path_type(),
			'admin' => '1',
		);
		$a = shortcode_atts( $defaults, $atts );

		// Validate attributes
		if ( ! in_array( $a['type'], array_keys( $this->get_template_path_types() ) ) ) {
			$a['type'] = $this->get_default_template_path_type();
		}

		$args = array(
			'admin_only' => ( '0' === $a['admin'] ? false : true ),
			'echo'       => false,
			'return'     => false,
		);

		return $this->reveal( $a['type'], $args );
	}

	/**
	 * Invokes the reveal intended to be shown in the site's footer.
	 *
	 * @since 3.0
	 */
	public function reveal_in_footer() {
		$options = $this->get_options();
		return $this->reveal( $options['template_path'], array( 'format_from_settings' => true ) );
	}

	/**
	 * Determines if the admin bar entry should be shown.
	 *
	 * @since 3.5
	 *
	 * @return bool True if the admin bar entry can be shown, else false.
	 */
	public function can_show_in_admin_bar() {
		// Don't show if in admin or is admin bar isn't even showing.
		if ( is_admin() || ! is_admin_bar_showing() ) {
			return false;
		}

		$options = $this->get_options();

		// Is feature enabled via settings and user is permitted to see it?
		return $options['display_in_admin_bar'] && $this->reveal_to_current_user();
	}

	/**
	 * Outputs CSS for the admin bar menu item.
	 *
	 * @since 3.5
	 */
	public function output_admin_bar_styles() {
		echo '<style>#wpadminbar #wp-admin-bar-reveal-template .ab-icon::before { content: "\f100"; top: 2px; } </style>' . "\n";
	}

	/**
	 * Adds an admin menu bar node that reveals the template.
	 *
	 * @since 3.5
	 *
	 * @param WP_Admin_Bar $wp_admin_bar WP_Admin_Bar instance.
	 */
	public function add_to_admin_bar( $wp_admin_bar ) {
		$wp_admin_bar->add_node(
			array(
				'id'     => 'reveal-template',
				'parent' => 'top-secondary',
				'title'  => '<span class="ab-icon"></span><span class="ab-label">' . __( 'Reveal Template', 'reveal-template' ) . '</span>',
				'meta'   => array(
					'tabindex' => 0,
				),
			)
		);

		$options = $this->get_options();
		$path = $this->reveal( $options['template_path'], array( 'echo' => false, 'format_from_settings' => false ) );

		$wp_admin_bar->add_node(
			array(
				'parent' => 'reveal-template',
				'id'     => 'revealed-template',
				'title'  => $path,
			)
		);
	}

	/**
	 * Formats for output the template path info for the currently rendered template.
	 *
	 * @param string $template_path_type The style of the template's path for return. Accepts: 'absolute', 'relative',
	 *                                   'theme-relative', 'filename'
	 * @param array  $args               {
	 *     Optional. Additional arguments.
	 *
	 *     @type bool   $admin_only           Only show for an admin? Default is true.
	 *     @type bool   $echo                 Echo the output? Default is true.
	 *     @type string $format               Output string format. Uses '%template%' as placeholder for template path.
	 *     @type bool   $format_from_settings Use format string specified via the plugin's settings page? Default false.
	 *     @type bool   $return               Return the value regardless of the admin_only value and check? Default true.
	 * }
	 * @return string   The path info for the currently rendered template, unless $args['return'] is false AND user
	 *                  wouldn't be shown the output
	 */
	public function reveal( $template_path_type, $args = array() ) {
		$template = $this->template;

		if ( ! $template ) {
			return '';
		}

		$defaults = array(
			'admin_only'           => true,
			'echo'                 => true,
			'format'               => '',
			'format_from_settings' => false,
			'return'               => true,
		);
		$args = wp_parse_args( $args, $defaults );

		$return = $args['return'];

		// Format the template path string based on type.
		switch ( $template_path_type ) {
			case 'absolute':
				// Do nothing; already have the absolute path
				break;
			case 'filename':
				$template = basename( $template );
				break;
			case 'relative':
				$template = str_replace( ABSPATH, '', $template );
				break;
			case 'theme-relative':
			default:
				$template = basename( dirname( $template ) ) . '/' . basename( $template );
				break;
		}

		// Check if current user can be shown the template.
		$is_allowed = ( false === $args['admin_only'] || $this->reveal_to_current_user() );

		if ( $return || $is_allowed ) {
			// Format the template path string according to format string.
			if ( $args['format_from_settings'] ) {
				$options = $this->get_options();
				$format = $options['format'];
			} else {
				$format = $args['format'];
			}
			$display = $format ? str_replace( '%template%', $template, $format ) : $template;

			if ( $is_allowed && $display && $args['echo'] ) {
				echo $display;
			}

			return $display;
		}

		return '';
	}

} // end c2c_RevealTemplate

add_action( 'plugins_loaded', array( 'c2c_RevealTemplate', 'get_instance' ) );

//
// TEMPLATE FUNCTION
//

	/**
	 * Formats the template path info for the currently rendered template for output.
	 *
	 * If $template_path_type argument is not specified, then the default value
	 * configured via the plugin's settings page will be used.
	 *
	 * @since 2.0
	 *
	 * @param bool   $echo               Optional. Echo the template info? Default true
	 * @param string $template_path_type Optional. The style of the template's path for return. Accepts: 'absolute',
	 *                                   'relative', 'theme-relative', 'filename'
	 * @param array  $args               Optional. Additional configuration. See c2c_RevealTemplate::reveal() for
	 *                                   documentation.
	 * @return string The path info for the currently rendered template
	 */
	if ( ! function_exists( 'c2c_reveal_template' ) ) :
		function c2c_reveal_template( $echo = true, $template_path_type = '', $args = array() ) {
			// See (and possibly override) 'echo' value in $args with value passed as the $echo argument.
			$args['echo'] = $echo;
			return c2c_RevealTemplate::get_instance()->reveal( $template_path_type, $args );
		}
	endif;

endif; // end if !class_exists()
